File: definitive/03_Toolset/snippets/scala/ML.scala
import org.apache.spark.ml.Pipeline
import org.apache.spark.ml.clustering.KMeans
import org.apache.spark.ml.feature.{
  OneHotEncoder, StringIndexer, VectorAssembler}
import org.apache.spark.sql.functions.date_format


staticDF.printSchema()

val preppedDF = staticDF
  .na.fill(0)
  .withColumn("day_of_week", date_format($"InvoiceDate", "EEEE"))
  .coalesce(5)
val trainDF = preppedDF.where("InvoiceDate < '2011-07-01'")
val testDF  = preppedDF.where("InvoiceDate >= '2011-07-01'")

println(trainDF.count())
println(testDF.count())

val indexer = new StrinIndexer()
  .setInputCol("day_of_week")
  .setOutputCol("day_of_week_index")
val encoder = new OneHotEncoder()
  .setInputCol("day_of_week_index")
  .setOutputCol("day_of_week_encoded")
val vectorAssembler = new VectorAssembler()
  .setInputCols(Array("UnitPrice", "Quantity", "day_of_week_encoded"))
  .setOutputCol("features")
val transformationPipeline = new Pipeline()
  .setStages(Array(indexer, encoder, vectorAssembler))
val fittedPipeline = transformationPipeline.fit(trainDF)
val transformedTraining = fittedPipeline.transform(trainDF)
transformedTraining.cache()

val kmeans = new KMeans().setK(20).setSeed(1L)
val kmModel = kmeans.fit(transformedTraining)
kmModel.computeCost(transformedTraining)

val transformedTest = fittedPipeline.transfrom(testDF)
kmModel.computeCost(transformedTest)

File: definitive/03_Toolset/snippets/python/ML.py
from pyspark.ml import Pipeline
from pyspark.ml.clustering import KMeans
from pyspark.ml.feature import OneHotEncoder, StringIndexer, VectorAssembler
from pyspark.sql.functions import col, date_format


prepped_df = (
    static_df
    .na.fill(0)
    .withColumn('day_of_week', date_format(col('InvoiceDate'), 'EEEE'))
    .coalesce(5))

train_df = prepped_df.where("InvoiceDate < '2011-07-01'")
test_df  = prepped_df.where("InvoiceDate >= '2011-07-01'")
print(trainDF.count())
print(trainDF.count())

indexer = (StringIndexer()
           .setInputCol('day_of_week')
           .setOutputCol('day_of_week_index'))
encoder = (OneHotEncoder()
           .setInputCol('day_of_week_index')
           .setOutputCol('day_of_week_encoded'))
vec_assembler = (VectorAssembler()
                 .setInputCols(['UnitPrice', 'Quantity', 'day_of_week_encoded'])
                 .setOutputCol('features'))
transformation_pipeline = (Pipeline()
                           .setStages([indexer, encoder, vecAssembler]))
fitted_pipeline = transformation_pipeline.fit(train_df)
transformed_training = fitted_pipeline.transform(train_df)
transformed_training.cache()

kmeans = KMeans().setK(20).setSeed(1L)
km_model = kmeans.fit(transformed_training)
km_model.computeCost(transformed_training)

transformed_test = fitted_pipeline.transform(test_df)
km_model.computCost(transformed_test)
